GitCommands/PathUtil.cs

```diff
@@ -124,14 +124,67 @@ public static string NormalizePath([NotNull] this string path)
 
             try
             {
-                return Path.GetFullPath(new Uri(path).LocalPath);
+                return Path.GetFullPath(Resolve(path));
             }
             catch (UriFormatException)
             {
                 return string.Empty;
             }
         }
 
+        [NotNull]
+        public static string Resolve([NotNull] string path, string relativePath = "")
+        {
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                throw new ArgumentException(nameof(path));
+            }
+
+            return path.IsWslPath() ? ResolveWsl(path, relativePath) : path.ResolveRelativePath(relativePath);
+        }
+
+        /// <summary>
+        /// Special handling of on purpose invalid WSL machine name in Windows 10
+        /// </summary>
+        [NotNull]
+        public static string ResolveWsl([NotNull] string path, string relativePath = "")
+        {
+            if (string.IsNullOrWhiteSpace(path) || !path.IsWslPath())
+            {
+                throw new ArgumentException(nameof(path));
+            }
+
+            // Temporarily replace machine name with a valid name (remove $ sign from \\wsl$\)
+            path = path.Remove(5, 1);
+
+            path = path.ResolveRelativePath(relativePath);
+
+            // Revert temporary replacement of WSL machine name (add $ sign back)
+            return path.Insert(5, "$");
+        }
+
+        [NotNull]
+        private static string ResolveRelativePath([NotNull] this string path, string relativePath)
+        {
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                throw new ArgumentException(nameof(path));
+            }
+
+            Uri tempPath = new Uri(path);
+            if (!string.IsNullOrEmpty(relativePath))
+            {
+                tempPath = new Uri(tempPath, relativePath);
+            }
+
+            return tempPath.LocalPath;
+        }
+
+        public static bool IsWslPath([NotNull] this string path)
```

Comment:
Doesn't need to be `public`, it is not used outside this class. For testing purposes you can make it `internal`
Equally doesn't need to be an extension.